flocker_bb/builders/maint.py

```diff
@@ -25,6 +43,123 @@ def makeCleanOldBuildsFactory():
+    :param prefixes: List of prefixes of instances to destroy.
+    """
+    name = 'clean-acceptance-instances'
+    description = ['Cleaning', 'acceptance', 'instances']
+    descriptionDone = ['Clean', 'acceptance', 'instances']
+    haltOnFailure = True
+    flunkOnFailure = True
+
+    def _thd_clean_nodes(self, config):
+        # Get the libcloud drivers corresponding to the acceptance tests.
+        config = yaml.safe_load(config)
+        rackspace = get_driver(Provider.RACKSPACE)(
+            config['rackspace']['username'], config['rackspace']['key'],
+            region=config['rackspace']['region'])
+        ec2 = get_driver(Provider.EC2)(
+            config['aws']['access_key'], config['aws']['secret_access_token'],
+            region=config['aws']['region'])
+
+        drivers = [rackspace, ec2]
+
+        # Get the prefixes of the instance names, appending the creator from
+        # the config.
+        creator = config['metadata']['creator']
+        prefixes = tuple(map(lambda prefix: prefix + creator, self.prefixes))
+
+        # Find out the cutoff time to use.
+        now = datetime.now(tzutc())
+        cutoff = now - self.lag
+
+        # Get all the nodes from the cloud providers
+        all_nodes = []
+        for driver in drivers:
+            all_nodes += driver.list_nodes()
+
+        # Filter them for running nodes with the right prefix.
+        test_nodes = [
+            node for node
+            in all_nodes
+            if node.name.startswith(prefixes)
+            and (node.state == NodeState.RUNNING)
+        ]
+
+        # Split the nodes into kept and destroyed nodes;
+        # destroying the ones older than the cut-off.
+        destroyed_nodes = []
+        kept_nodes = []
+        for node in test_nodes:
+            creation_time = get_creation_time(node)
+            if creation_time is not None and creation_time < cutoff:
+                node.destroy()
+                destroyed_nodes.append(node)
+            else:
+                kept_nodes.append(node)
+
+        return {
+            'destroyed': destroyed_nodes,
+            'kept': kept_nodes,
+        }
+
+    def start(self):
+        config = privateData['acceptance']['config']
+        d = deferToThread(self._thd_clean_nodes, config)
+        d.addCallback(self.log)
+        d.addErrback(self.failed)
+
+    def log(self, result):
+        """
+        Log the nodes kept and destroyed.
+        """
+        for kind, nodes in result.iteritems():
+            content = json.dumps([
+                {
+                    'id': node.id,
+                    'name': node.name,
+                    'provider': node.driver.name,
+                    'creatation_time': get_creation_time(node).isoformat(),
+                }
+                for node in nodes
+            ], sort_keys=True, indent=4, separators=(',', ': '))
+            self.addCompleteLog(name=kind, text=content)
+        self.finished(SUCCESS)
```

Comment:
> what happens when the buildmaster is restarted, will those slaves be lost in the wild ? maybe it would be better to have the nodes kill themselves, instead of having the buildmaster terminating them

This is the code that makes sure the acceptance test nodes (not slaves) aren't lost in the wild for too long.